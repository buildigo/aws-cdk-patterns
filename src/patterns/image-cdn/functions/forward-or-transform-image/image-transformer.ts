import {Format} from '../image-api.types'
import Sharp from 'sharp'
import {ImageTransformer} from './image-transformer.types'

export class SharpImageTransformer implements ImageTransformer {
  transform(image: Buffer, parameters: {width?: number; height?: number; format?: Format}): Promise<Buffer> {
    let stream = Sharp(image)
    if (parameters.height || parameters.width) {
      stream = stream.resize({width: parameters.width, height: parameters.height})
    }
    if (parameters.format) {
      switch (parameters.format) {
        case Format.JPG:
          stream = stream.toFormat('jpeg')
          break
        case Format.PNG:
          stream = stream.toFormat('png')
          break
        case Format.WEBP:
          stream = stream.toFormat('webp')
          break
        case Format.AVIF:
          stream = stream.toFormat('avif', {effort: 2, chromaSubsampling: '4:2:0'})
          break
      }
    }
    return stream.toBuffer()
  }
}
