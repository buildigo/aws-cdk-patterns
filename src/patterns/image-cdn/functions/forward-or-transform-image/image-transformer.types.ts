import {Format} from '../image-api.types'

export interface ImageTransformer {
  transform(image: Buffer, parameters: {width?: number; height?: number; format?: Format}): Promise<Buffer>
}
