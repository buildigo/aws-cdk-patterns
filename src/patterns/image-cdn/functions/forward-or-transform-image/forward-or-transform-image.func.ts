import {S3Client} from '@aws-sdk/client-s3'
import {CloudFrontResponseEvent, CloudFrontResponseResult, Handler} from 'aws-lambda'
import {SharpImageTransformer} from './image-transformer'
import {processEvent} from './forward-or-transform-image.handler'

const s3 = new S3Client({})
const imageTransformer = new SharpImageTransformer()

export const handler: Handler<CloudFrontResponseEvent, CloudFrontResponseResult> = event =>
  processEvent(s3, imageTransformer, event)
