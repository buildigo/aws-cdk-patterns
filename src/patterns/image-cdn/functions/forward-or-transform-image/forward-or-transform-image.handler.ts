import {GetObjectCommand, PutObjectCommand, S3Client} from '@aws-sdk/client-s3'
import {ImageTransformer} from './image-transformer.types'
import {CloudFrontResponseEvent, CloudFrontResponseResult} from 'aws-lambda'
import {parse} from 'querystring'
import {Format, IMAGE_API_PARAMS} from '../image-api.types'
import {isString} from 'lodash'
import {parseDimensionValue} from '../common/dimension'

export const processEvent = async (
  s3: S3Client,
  imageTransformer: ImageTransformer,
  event: CloudFrontResponseEvent,
): Promise<CloudFrontResponseResult> => {
  const {
    cf: {
      request: {querystring, uri},
      response,
      response: {status},
    },
  } = event.Records[0]

  console.debug('got event !', JSON.stringify(event))

  const bucketName = getOriginBucketName(event)
  // non s3 origin, nothing to do
  if (!bucketName) {
    console.warn('could not find s3 bucket name, returning...')
    return response
  }

  // image was found or any other error, nothing to do
  if (!['403', '404'].includes(status)) {
    return response
  }

  const queryParams = parse(querystring)
  const sourceImage = queryParams[IMAGE_API_PARAMS.SRC_IMAGE] as string
  const nextExtension = queryParams[IMAGE_API_PARAMS.EXTENSION] as string
  const w = queryParams[IMAGE_API_PARAMS.WIDTH]
  const h = queryParams[IMAGE_API_PARAMS.HEIGHT]
  const fmt = queryParams[IMAGE_API_PARAMS.FORMAT]

  if (!isString(sourceImage)) {
    console.warn('source image is not a string, returning', sourceImage)
    return response
  }
  const sourceKey = sourceImage.replace(/^\//, '')
  const width = parseDimensionValue(w)
  const height = parseDimensionValue(h)
  const format = parseFormat(fmt)

  if (!width && !height && !format) {
    return response
  }
  try {
    const contentType = mimeType(nextExtension)
    const key = uri.replace(/^\//, '')

    console.debug(`Querying source image: ${sourceKey} in bucket ${bucketName}`)
    const command = new GetObjectCommand({Bucket: bucketName, Key: sourceKey})
    const imageObj = await s3.send(command)
    const data = await imageObj.Body?.transformToByteArray()
    if (!data) {
      console.warn('Could not get object data', imageObj.Body)
      return response
    }
    const resizedImage = await imageTransformer.transform(
      Buffer.from(data),
      {width, format, height},
      {withoutEnlargement: true},
    )

    // TODO: set cache control on original image, and use it here
    const cacheControl = 'max-age=31536000,public,immutable'
    await s3.send(
      new PutObjectCommand({
        Body: resizedImage,
        Bucket: bucketName,
        ContentType: contentType,
        Key: key,
        StorageClass: 'STANDARD',
        CacheControl: cacheControl,
      }),
    )

    return {
      ...response,
      status: '200',
      statusDescription: 'Found',
      body: resizedImage.toString('base64'),
      bodyEncoding: 'base64',
      headers: {
        ...response.headers,
        'content-type': [{key: 'Content-Type', value: contentType}],
        'cache-control': [{key: 'Cache-Control', value: cacheControl}],
      },
    }
  } catch (error: unknown) {
    const errorMessage = `Error while getting source image object "${sourceKey}": ${error}`
    console.error('Error caught', error)

    return {
      ...response,
      status: '404',
      statusDescription: 'Not Found',
      body: errorMessage,
      bodyEncoding: 'text',
      headers: {
        ...response.headers,
        'content-type': [{key: 'Content-Type', value: 'text/plain'}],
      },
    }
  }
}

function getOriginBucketName(event: CloudFrontResponseEvent): string | undefined {
  const {
    cf: {
      request: {origin},
    },
  } = event.Records[0]
  if (origin?.s3) {
    const domainName = origin.s3.domainName
    return parseS3BucketName(domainName)
  } else if (origin?.custom) {
    const domainName = origin.custom.domainName
    if (!domainName) {
      console.warn('could not find custom origin domain name, returning ...')
      return undefined
    } else if (!domainName.includes('s3-website')) {
      console.warn('custom origin is not an s3-website (unsupported), returning ...')
      return undefined
    } else {
      return parseS3BucketName(domainName)
    }
  } else {
    return undefined
  }
}

export function parseS3BucketName(domainName: string): string | undefined {
  const bucketNameRegex = /[^/]+(?=.(?:s3|s3-website)\.(?:[a-z0-9-]+)\.amazonaws\.com)/i
  const matches = domainName.match(bucketNameRegex)
  if (matches?.length) {
    return matches[0]
  } else {
    return undefined
  }
}

function parseFormat(fmt?: unknown): Format | undefined {
  if (!isString(fmt)) {
    return undefined
  }
  switch (fmt) {
    case 'jpeg':
      return Format.JPG
    case 'png':
      return Format.PNG
    case 'avif':
      return Format.AVIF
    case 'webp':
      return Format.WEBP
    default:
      return undefined
  }
}

function mimeType(extension: string): string {
  switch (extension) {
    case 'jpg':
    case 'jpeg':
      return 'image/jpeg'
    case 'avif':
      return 'image/avif'
    case 'webp':
      return 'image/webp'
    case 'png':
      return 'image/png'
    default:
      console.warn(`Could not determine mime type for extension: '${extension}'`)
      return ''
  }
}
