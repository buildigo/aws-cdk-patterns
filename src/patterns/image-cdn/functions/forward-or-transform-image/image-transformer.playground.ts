import {SharpImageTransformer} from './image-transformer'
import {readFile, writeFile} from 'fs/promises'
import {Format} from '../image-api.types'
import {randomUUID} from 'crypto'

async function main() {
  const t = new SharpImageTransformer()
  const image = await readFile('test.jpg')
  const format = Format.AVIF
  const res = await t.transform(image, {height: 300, format})
  await writeFile(`${randomUUID()}.${format.toLowerCase()}`, res)
}

main().then(() => console.log('done'))
