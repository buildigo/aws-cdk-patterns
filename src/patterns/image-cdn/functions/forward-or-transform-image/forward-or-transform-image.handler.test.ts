import {parseS3BucketName, processEvent as ForwardOrTransformImage} from './forward-or-transform-image.handler'
import {CloudFrontResponseEvent} from 'aws-lambda'
import {GetObjectCommand, S3Client} from '@aws-sdk/client-s3'
import {ImageTransformer} from './image-transformer.types'
import {Format} from '../image-api.types'
import {sdkStreamMixin} from '@aws-sdk/util-stream-node'
import {Readable} from 'stream'
import {mockS3Client} from '../../../../../test/s3.mock'

describe('ForwardOrTransformImage', () => {
  beforeEach(() => {
    mockS3Responses()
  })
  it('should return the resized image if it already exists', async () => {
    const response = await ForwardOrTransformImage(new S3Client({}), mockImageTransformer(), mock200Event())
    expect(response!.status).toBe('200')
    expect(response).toMatchSnapshot()
  })

  it('should resize the image when it does not exist yet (403), but the source image does exist', async () => {
    const response = await ForwardOrTransformImage(new S3Client({}), mockImageTransformer(), mock403Event())
    expect(response!.status).toBe('200')
    expect(response!.headers!['content-type'][0].value).toBe('image/webp')
    expect(response).toMatchSnapshot()
  })

  it('should resize the image when it does not exist (404), but the source image does exist', async () => {
    const response = await ForwardOrTransformImage(new S3Client({}), mockImageTransformer(), mock404Event())
    expect(response!.status).toBe('200')
    expect(response!.headers!['content-type'][0].value).toBe('image/webp')
    expect(response).toMatchSnapshot()
  })

  it('should return 404 when source image does not exist', async () => {
    const response = await ForwardOrTransformImage(new S3Client({}), mockImageTransformer(), mockBadKeyEvent())
    expect(response).toMatchSnapshot()
  })

  it('should support events from a bucket with static website enabled (custom origin)', async () => {
    const response = await ForwardOrTransformImage(
      new S3Client({}),
      mockImageTransformer(),
      mockStaticWebsite404Event(),
    )
    expect(response!.status).toBe('200')
    expect(response!.headers!['content-type'][0].value).toBe('image/avif')

    expect(response).toMatchSnapshot()
  })
})

describe('parseS3BucketName', () => {
  const tests: [string, string | undefined][] = [
    ['https://image-test-yeah.s3-website.eu-central-1.amazonaws.com', 'image-test-yeah'],
    ['https://image-test-images.s3.eu-central-1.amazonaws.com', 'image-test-images'],
    ['https://image-test-images-xxx.s3.us-east-1.amazonaws.com', 'image-test-images-xxx'],
    ['image-test-images-xxx.s3.us-east-1.amazonaws.com', 'image-test-images-xxx'],
    ['abcd.image-test-images-xxx.us-east-1.amazonaws.com', undefined],
    ['test.com', undefined],
  ]

  for (const test of tests) {
    const [url, bucketName] = test
    it(`matches the bucket name "${bucketName}" from "${url}"`, () => {
      expect(parseS3BucketName(url)).toEqual(bucketName)
    })
  }
})

const mock200Event = (
  uri = 'good-image.png',
  querystring = 'ext=webp&fmt=webp&height=400&sourceImage=/good-image.png&width=700',
): CloudFrontResponseEvent => ({
  Records: [
    {
      cf: {
        config: {
          distributionDomainName: 'd1ppm5fhf7dvj2.cloudfront.net',
          distributionId: 'E3FLSY2KR0CQYV',
          eventType: 'origin-response',
          requestId: 'ZglsXKNU9HC_gwfO1qhXqGh_AipNOvr73VwCe0CTN3-Guu-b1C5MEw==',
        },
        request: {
          clientIp: '108.41.158.64',
          headers: {
            'x-forwarded-for': [
              {
                key: 'X-Forwarded-For',
                value: '108.41.158.64',
              },
            ],
            'user-agent': [
              {
                key: 'User-Agent',
                value: 'Amazon CloudFront',
              },
            ],
            via: [
              {
                key: 'Via',
                value: '1.1 bcf69805bd7618b134f85cb9528fba32.cloudfront.net (CloudFront)',
              },
            ],
            'upgrade-insecure-requests': [
              {
                key: 'Upgrade-Insecure-Requests',
                value: '1',
              },
            ],
            'accept-encoding': [
              {
                key: 'Accept-Encoding',
                value: 'gzip',
              },
            ],
            host: [
              {
                key: 'Host',
                value: 'image-test-images.s3.amazonaws.com',
              },
            ],
            'cache-control': [
              {
                key: 'Cache-Control',
                value: 'max-age=0',
              },
            ],
          },
          method: 'GET',
          origin: {
            s3: {
              authMethod: 'origin-access-identity',
              customHeaders: {},
              domainName: 'image-test-images.s3.eu-central-1.amazonaws.com',
              path: '',
              region: 'eu-central-1',
            },
          },
          querystring,
          uri,
        },
        response: {
          headers: {
            'x-amz-id-2': [
              {
                key: 'x-amz-id-2',
                value: '8q8ejgujau4YCUdFL3Ck1VPB6PsiEr/60RIZPUDxgOuzM/t/Dueg5SG1RdbKxnFKN7u+Yt8QkiI=',
              },
            ],
            'x-amz-request-id': [
              {
                key: 'x-amz-request-id',
                value: '37THKNX0N2ADYKS3',
              },
            ],
            date: [
              {
                key: 'Date',
                value: 'Sat, 08 May 2021 23:36:25 GMT',
              },
            ],
            'last-modified': [
              {
                key: 'Last-Modified',
                value: 'Sat, 08 May 2021 23:33:28 GMT',
              },
            ],
            etag: [
              {
                key: 'ETag',
                value: '"72a9ff667c12b7d156a696877fc4df89"',
              },
            ],
            'accept-ranges': [
              {
                key: 'Accept-Ranges',
                value: 'bytes',
              },
            ],
            server: [
              {
                key: 'Server',
                value: 'AmazonS3',
              },
            ],
            'content-type': [
              {
                key: 'Content-Type',
                value: 'image/webp',
              },
            ],
            'content-length': [
              {
                key: 'Content-Length',
                value: '2832',
              },
            ],
          },
          status: '200',
          statusDescription: 'OK',
        },
      },
    },
  ],
})

const mock403Event = (
  uri = 'good-image.png',
  querystring = 'ext=webp&fmt=webp&height=400&sourceImage=/good-image.png&width=700',
): CloudFrontResponseEvent => ({
  Records: [
    {
      cf: {
        config: {
          distributionDomainName: 'd1ppm5fhf7dvj2.cloudfront.net',
          distributionId: 'E3FLSY2KR0CQYV',
          eventType: 'origin-response',
          requestId: '6bog7lAuZZ6TdQmXtL9RuwBqgvRN9Ok2ehhJbJ_jTkQOY1mRt3WIrQ==',
        },
        request: {
          clientIp: '108.41.158.64',
          headers: {
            'x-forwarded-for': [
              {
                key: 'X-Forwarded-For',
                value: '108.41.158.64',
              },
            ],
            'user-agent': [
              {
                key: 'User-Agent',
                value: 'Amazon CloudFront',
              },
            ],
            via: [
              {
                key: 'Via',
                value: '1.1 bcf69805bd7618b134f85cb9528fba32.cloudfront.net (CloudFront)',
              },
            ],
            'upgrade-insecure-requests': [
              {
                key: 'Upgrade-Insecure-Requests',
                value: '1',
              },
            ],
            'accept-encoding': [
              {
                key: 'Accept-Encoding',
                value: 'gzip',
              },
            ],
            host: [
              {
                key: 'Host',
                value: 'image-test-images.s3.amazonaws.com',
              },
            ],
            'cache-control': [
              {
                key: 'Cache-Control',
                value: 'max-age=0',
              },
            ],
          },
          method: 'GET',
          origin: {
            s3: {
              authMethod: 'origin-access-identity',
              customHeaders: {},
              domainName: 'image-test-images.s3.eu-central-1.amazonaws.com',
              path: '',
              region: 'eu-central-1',
            },
          },
          querystring,
          uri,
        },
        response: {
          headers: {
            'x-amz-request-id': [
              {
                key: 'x-amz-request-id',
                value: 'XPWJPWV1KP5CFDTS',
              },
            ],
            'x-amz-id-2': [
              {
                key: 'x-amz-id-2',
                value: 'QZlgZ/jWFGkc/pjNjijOkI1KjEH7avdgXL/2r5kuD7xMWFs6K3CSnHycKhzcVrN1chM0Hkq+IQM=',
              },
            ],
            date: [
              {
                key: 'Date',
                value: 'Sat, 08 May 2021 23:33:25 GMT',
              },
            ],
            server: [
              {
                key: 'Server',
                value: 'AmazonS3',
              },
            ],
            'content-type': [
              {
                key: 'Content-Type',
                value: 'application/xml',
              },
            ],
            'transfer-encoding': [
              {
                key: 'Transfer-Encoding',
                value: 'chunked',
              },
            ],
          },
          status: '403',
          statusDescription: 'Forbidden',
        },
      },
    },
  ],
})

const mock404Event = (
  uri = 'good-image.png',
  querystring = 'ext=webp&fmt=webp&height=400&sourceImage=/good-image.png&width=700',
): CloudFrontResponseEvent => ({
  Records: [
    {
      cf: {
        config: {
          distributionDomainName: 'd1ppm5fhf7dvj2.cloudfront.net',
          distributionId: 'E3FLSY2KR0CQYV',
          eventType: 'origin-response',
          requestId: '6bog7lAuZZ6TdQmXtL9RuwBqgvRN9Ok2ehhJbJ_jTkQOY1mRt3WIrQ==',
        },
        request: {
          clientIp: '108.41.158.64',
          headers: {
            'x-forwarded-for': [
              {
                key: 'X-Forwarded-For',
                value: '108.41.158.64',
              },
            ],
            'user-agent': [
              {
                key: 'User-Agent',
                value: 'Amazon CloudFront',
              },
            ],
            via: [
              {
                key: 'Via',
                value: '1.1 bcf69805bd7618b134f85cb9528fba32.cloudfront.net (CloudFront)',
              },
            ],
            'upgrade-insecure-requests': [
              {
                key: 'Upgrade-Insecure-Requests',
                value: '1',
              },
            ],
            'accept-encoding': [
              {
                key: 'Accept-Encoding',
                value: 'gzip',
              },
            ],
            host: [
              {
                key: 'Host',
                value: 'image-test-images.s3.eu-central-1.amazonaws.com',
              },
            ],
            'cache-control': [
              {
                key: 'Cache-Control',
                value: 'max-age=0',
              },
            ],
          },
          method: 'GET',
          origin: {
            s3: {
              authMethod: 'origin-access-identity',
              customHeaders: {},
              domainName: 'image-test-images.s3.eu-central-1.amazonaws.com',
              path: '',
              region: 'eu-central-1',
            },
          },
          querystring,
          uri,
        },
        response: {
          headers: {
            'x-amz-request-id': [
              {
                key: 'x-amz-request-id',
                value: 'XPWJPWV1KP5CFDTS',
              },
            ],
            'x-amz-id-2': [
              {
                key: 'x-amz-id-2',
                value: 'QZlgZ/jWFGkc/pjNjijOkI1KjEH7avdgXL/2r5kuD7xMWFs6K3CSnHycKhzcVrN1chM0Hkq+IQM=',
              },
            ],
            date: [
              {
                key: 'Date',
                value: 'Sat, 08 May 2021 23:33:25 GMT',
              },
            ],
            server: [
              {
                key: 'Server',
                value: 'AmazonS3',
              },
            ],
            'content-type': [
              {
                key: 'Content-Type',
                value: 'application/xml',
              },
            ],
            'transfer-encoding': [
              {
                key: 'Transfer-Encoding',
                value: 'chunked',
              },
            ],
          },
          status: '404',
          statusDescription: 'Not Found',
        },
      },
    },
  ],
})

const mockBadKeyEvent = (
  uri = 'bad-image.png',
  querystring = 'ext=webp&fmt=webp&height=400&sourceImage=/bad-image.png&width=700',
): CloudFrontResponseEvent => ({
  Records: [
    {
      cf: {
        config: {
          distributionDomainName: 'd1ppm5fhf7dvj2.cloudfront.net',
          distributionId: 'E3FLSY2KR0CQYV',
          eventType: 'origin-response',
          requestId: 'KN-wBwbmbE62plUp5mxC6hxzoXSOt9n6OZ06z2LjULGpbpOCpRWwqw==',
        },
        request: {
          clientIp: '108.41.158.64',
          headers: {
            'x-forwarded-for': [
              {
                key: 'X-Forwarded-For',
                value: '108.41.158.64',
              },
            ],
            'user-agent': [
              {
                key: 'User-Agent',
                value: 'Amazon CloudFront',
              },
            ],
            via: [
              {
                key: 'Via',
                value: '1.1 4d8d3505beb1aa504be6a19a0d2e784c.cloudfront.net (CloudFront)',
              },
            ],
            'upgrade-insecure-requests': [
              {
                key: 'Upgrade-Insecure-Requests',
                value: '1',
              },
            ],
            'accept-encoding': [
              {
                key: 'Accept-Encoding',
                value: 'gzip',
              },
            ],
            host: [
              {
                key: 'Host',
                value: 'image-test-images.s3.amazonaws.com',
              },
            ],
          },
          method: 'GET',
          origin: {
            s3: {
              authMethod: 'origin-access-identity',
              customHeaders: {},
              domainName: 'image-test-images.s3.eu-central-1.amazonaws.com',
              path: '',
              region: 'eu-central-1',
            },
          },
          querystring,
          uri,
        },
        response: {
          headers: {
            'x-amz-request-id': [
              {
                key: 'x-amz-request-id',
                value: 'VWT4ZVZW6KVSZXK5',
              },
            ],
            'x-amz-id-2': [
              {
                key: 'x-amz-id-2',
                value: '4/PLziA+bP9cjFWVFErAyGd5NuOWY4E8XuE3DH0J913HugdpsWOCg+MsIAapBR8CXjT15VBywq8=',
              },
            ],
            date: [
              {
                key: 'Date',
                value: 'Sat, 08 May 2021 23:38:02 GMT',
              },
            ],
            server: [
              {
                key: 'Server',
                value: 'AmazonS3',
              },
            ],
            'content-type': [
              {
                key: 'Content-Type',
                value: 'application/xml',
              },
            ],
            'transfer-encoding': [
              {
                key: 'Transfer-Encoding',
                value: 'chunked',
              },
            ],
          },
          status: '404',
          statusDescription: 'Not Found',
        },
      },
    },
  ],
})

const mockStaticWebsite404Event = (
  uri = 'good-image.png',
  querystring = 'ext=avif&fmt=avif&height=200&sourceImage=/good-image.png&width=700',
): CloudFrontResponseEvent => ({
  Records: [
    {
      cf: {
        config: {
          distributionDomainName: 'd1ppm5fhf7dvj2.cloudfront.net',
          distributionId: 'E3FLSY2KR0CQYV',
          eventType: 'origin-response',
          requestId: 'ZglsXKNU9HC_gwfO1qhXqGh_AipNOvr73VwCe0CTN3-Guu-b1C5MEw==',
        },
        request: {
          clientIp: '108.41.158.64',
          headers: {
            'x-forwarded-for': [
              {
                key: 'X-Forwarded-For',
                value: '108.41.158.64',
              },
            ],
            'user-agent': [
              {
                key: 'User-Agent',
                value: 'Amazon CloudFront',
              },
            ],
            via: [
              {
                key: 'Via',
                value: '1.1 bcf69805bd7618b134f85cb9528fba32.cloudfront.net (CloudFront)',
              },
            ],
            'upgrade-insecure-requests': [
              {
                key: 'Upgrade-Insecure-Requests',
                value: '1',
              },
            ],
            'accept-encoding': [
              {
                key: 'Accept-Encoding',
                value: 'gzip',
              },
            ],
            host: [
              {
                key: 'Host',
                value: 'image-test-images.s3.amazonaws.com',
              },
            ],
            'cache-control': [
              {
                key: 'Cache-Control',
                value: 'max-age=0',
              },
            ],
          },
          method: 'GET',
          origin: {
            custom: {
              customHeaders: {},
              domainName: 'image-test-yeah.s3-website.eu-central-1.amazonaws.com',
              keepaliveTimeout: 5,
              path: '',
              port: 80,
              protocol: 'http',
              readTimeout: 30,
              sslProtocols: ['TLSv1.2'],
            },
          },
          querystring,
          uri,
        },
        response: {
          headers: {
            'x-amz-id-2': [
              {
                key: 'x-amz-id-2',
                value: '8q8ejgujau4YCUdFL3Ck1VPB6PsiEr/60RIZPUDxgOuzM/t/Dueg5SG1RdbKxnFKN7u+Yt8QkiI=',
              },
            ],
            'x-amz-request-id': [
              {
                key: 'x-amz-request-id',
                value: '37THKNX0N2ADYKS3',
              },
            ],
            date: [
              {
                key: 'Date',
                value: 'Sat, 08 May 2021 23:36:25 GMT',
              },
            ],
            'last-modified': [
              {
                key: 'Last-Modified',
                value: 'Sat, 08 May 2021 23:33:28 GMT',
              },
            ],
            etag: [
              {
                key: 'ETag',
                value: '"72a9ff667c12b7d156a696877fc4df89"',
              },
            ],
            'accept-ranges': [
              {
                key: 'Accept-Ranges',
                value: 'bytes',
              },
            ],
            server: [
              {
                key: 'Server',
                value: 'AmazonS3',
              },
            ],
            'content-type': [
              {
                key: 'Content-Type',
                value: 'application/xml',
              },
            ],
            'content-length': [
              {
                key: 'Content-Length',
                value: '2832',
              },
            ],
          },
          status: '404',
          statusDescription: 'Not Found',
        },
      },
    },
  ],
})

function mockS3Responses() {
  const mockImage =
    'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg=='
  const validObjectKey = 'good-image.png'

  return mockS3Client()
    .on(GetObjectCommand)
    .rejects(new Error(`Invalid object key`))
    .on(GetObjectCommand, {Key: validObjectKey})
    .resolves({
      Body: sdkStreamMixin(Readable.from(Buffer.from(mockImage, 'base64'))),
    })
}

const mockImageTransformer = (): ImageTransformer => ({
  transform: (image: Buffer, parameters: {width?: number; height?: number; format?: Format}): Promise<Buffer> => {
    return Promise.resolve(image)
  },
})
