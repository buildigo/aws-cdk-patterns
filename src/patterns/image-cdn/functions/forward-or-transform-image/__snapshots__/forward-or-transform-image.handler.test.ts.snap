// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForwardOrTransformImage should resize the image when it does not exist (404), but the source image does exist 1`] = `
Object {
  "body": "dataimage/pngbase64iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJgg",
  "bodyEncoding": "base64",
  "headers": Object {
    "cache-control": Array [
      Object {
        "key": "Cache-Control",
        "value": "max-age=31536000,public,immutable",
      },
    ],
    "content-type": Array [
      Object {
        "key": "Content-Type",
        "value": "image/webp",
      },
    ],
    "date": Array [
      Object {
        "key": "Date",
        "value": "Sat, 08 May 2021 23:33:25 GMT",
      },
    ],
    "server": Array [
      Object {
        "key": "Server",
        "value": "AmazonS3",
      },
    ],
    "transfer-encoding": Array [
      Object {
        "key": "Transfer-Encoding",
        "value": "chunked",
      },
    ],
    "x-amz-id-2": Array [
      Object {
        "key": "x-amz-id-2",
        "value": "QZlgZ/jWFGkc/pjNjijOkI1KjEH7avdgXL/2r5kuD7xMWFs6K3CSnHycKhzcVrN1chM0Hkq+IQM=",
      },
    ],
    "x-amz-request-id": Array [
      Object {
        "key": "x-amz-request-id",
        "value": "XPWJPWV1KP5CFDTS",
      },
    ],
  },
  "status": "200",
  "statusDescription": "Found",
}
`;

exports[`ForwardOrTransformImage should resize the image when it does not exist yet (403), but the source image does exist 1`] = `
Object {
  "body": "dataimage/pngbase64iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJgg",
  "bodyEncoding": "base64",
  "headers": Object {
    "cache-control": Array [
      Object {
        "key": "Cache-Control",
        "value": "max-age=31536000,public,immutable",
      },
    ],
    "content-type": Array [
      Object {
        "key": "Content-Type",
        "value": "image/webp",
      },
    ],
    "date": Array [
      Object {
        "key": "Date",
        "value": "Sat, 08 May 2021 23:33:25 GMT",
      },
    ],
    "server": Array [
      Object {
        "key": "Server",
        "value": "AmazonS3",
      },
    ],
    "transfer-encoding": Array [
      Object {
        "key": "Transfer-Encoding",
        "value": "chunked",
      },
    ],
    "x-amz-id-2": Array [
      Object {
        "key": "x-amz-id-2",
        "value": "QZlgZ/jWFGkc/pjNjijOkI1KjEH7avdgXL/2r5kuD7xMWFs6K3CSnHycKhzcVrN1chM0Hkq+IQM=",
      },
    ],
    "x-amz-request-id": Array [
      Object {
        "key": "x-amz-request-id",
        "value": "XPWJPWV1KP5CFDTS",
      },
    ],
  },
  "status": "200",
  "statusDescription": "Found",
}
`;

exports[`ForwardOrTransformImage should return 404 when source image does not exist 1`] = `
Object {
  "body": "Error while getting source image object \\"bad-image.png\\": Error: Invalid object key",
  "bodyEncoding": "text",
  "headers": Object {
    "content-type": Array [
      Object {
        "key": "Content-Type",
        "value": "text/plain",
      },
    ],
    "date": Array [
      Object {
        "key": "Date",
        "value": "Sat, 08 May 2021 23:38:02 GMT",
      },
    ],
    "server": Array [
      Object {
        "key": "Server",
        "value": "AmazonS3",
      },
    ],
    "transfer-encoding": Array [
      Object {
        "key": "Transfer-Encoding",
        "value": "chunked",
      },
    ],
    "x-amz-id-2": Array [
      Object {
        "key": "x-amz-id-2",
        "value": "4/PLziA+bP9cjFWVFErAyGd5NuOWY4E8XuE3DH0J913HugdpsWOCg+MsIAapBR8CXjT15VBywq8=",
      },
    ],
    "x-amz-request-id": Array [
      Object {
        "key": "x-amz-request-id",
        "value": "VWT4ZVZW6KVSZXK5",
      },
    ],
  },
  "status": "404",
  "statusDescription": "Not Found",
}
`;

exports[`ForwardOrTransformImage should return the resized image if it already exists 1`] = `
Object {
  "headers": Object {
    "accept-ranges": Array [
      Object {
        "key": "Accept-Ranges",
        "value": "bytes",
      },
    ],
    "content-length": Array [
      Object {
        "key": "Content-Length",
        "value": "2832",
      },
    ],
    "content-type": Array [
      Object {
        "key": "Content-Type",
        "value": "image/webp",
      },
    ],
    "date": Array [
      Object {
        "key": "Date",
        "value": "Sat, 08 May 2021 23:36:25 GMT",
      },
    ],
    "etag": Array [
      Object {
        "key": "ETag",
        "value": "\\"72a9ff667c12b7d156a696877fc4df89\\"",
      },
    ],
    "last-modified": Array [
      Object {
        "key": "Last-Modified",
        "value": "Sat, 08 May 2021 23:33:28 GMT",
      },
    ],
    "server": Array [
      Object {
        "key": "Server",
        "value": "AmazonS3",
      },
    ],
    "x-amz-id-2": Array [
      Object {
        "key": "x-amz-id-2",
        "value": "8q8ejgujau4YCUdFL3Ck1VPB6PsiEr/60RIZPUDxgOuzM/t/Dueg5SG1RdbKxnFKN7u+Yt8QkiI=",
      },
    ],
    "x-amz-request-id": Array [
      Object {
        "key": "x-amz-request-id",
        "value": "37THKNX0N2ADYKS3",
      },
    ],
  },
  "status": "200",
  "statusDescription": "OK",
}
`;

exports[`ForwardOrTransformImage should support events from a bucket with static website enabled (custom origin) 1`] = `
Object {
  "body": "dataimage/pngbase64iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJgg",
  "bodyEncoding": "base64",
  "headers": Object {
    "accept-ranges": Array [
      Object {
        "key": "Accept-Ranges",
        "value": "bytes",
      },
    ],
    "cache-control": Array [
      Object {
        "key": "Cache-Control",
        "value": "max-age=31536000,public,immutable",
      },
    ],
    "content-length": Array [
      Object {
        "key": "Content-Length",
        "value": "2832",
      },
    ],
    "content-type": Array [
      Object {
        "key": "Content-Type",
        "value": "image/avif",
      },
    ],
    "date": Array [
      Object {
        "key": "Date",
        "value": "Sat, 08 May 2021 23:36:25 GMT",
      },
    ],
    "etag": Array [
      Object {
        "key": "ETag",
        "value": "\\"72a9ff667c12b7d156a696877fc4df89\\"",
      },
    ],
    "last-modified": Array [
      Object {
        "key": "Last-Modified",
        "value": "Sat, 08 May 2021 23:33:28 GMT",
      },
    ],
    "server": Array [
      Object {
        "key": "Server",
        "value": "AmazonS3",
      },
    ],
    "x-amz-id-2": Array [
      Object {
        "key": "x-amz-id-2",
        "value": "8q8ejgujau4YCUdFL3Ck1VPB6PsiEr/60RIZPUDxgOuzM/t/Dueg5SG1RdbKxnFKN7u+Yt8QkiI=",
      },
    ],
    "x-amz-request-id": Array [
      Object {
        "key": "x-amz-request-id",
        "value": "37THKNX0N2ADYKS3",
      },
    ],
  },
  "status": "200",
  "statusDescription": "Found",
}
`;
